glue/core/data_factories/helpers.py

```diff
@@ -272,48 +267,52 @@ def data_label(path):
 
 @contract(extension='string', factory='callable')
 def set_default_factory(extension, factory):
-    """Register an extension that should be handled by a factory by default
-
-    :param extension: File extension (do not include the '.')
-    :param factory: The factory function to dispatch to
-    """
-    for ex in extension.split():
-        _default_factory[ex] = factory
+    warnings.warn("set_default_factory is deprecated and no longer has any effect")
 
 
 @contract(extension='string', returns='callable|None')
 def get_default_factory(extension):
-    """Return the default factory function to read a given file extension.
-
-    :param extension: The extension to lookup
-
-    :rtype: A factory function, or None if the extension has no default
-    """
-    try:
-        return _default_factory[extension]
-    except KeyError:
-        return None
+    warnings.warn("get_default_factory is deprecated and will always return None")
+    return None
 
 
 @contract(filename='string')
 def find_factory(filename, **kwargs):
+
     from ...config import data_factory
 
-    # on first pass, only try the default factory
-    default = _default_factory.get(_extension(filename))
-    for func, _, identifier in data_factory:
+    # We no longer try the 'default' factory first because we actually need to
+    # try all identifiers and select the one to use based on the priority. This
+    # allows us to define more specialized loaders take priority over more
+    # general ones. For example, a FITS file that is a dendrogram should be
+    # loaded as a dendrogram, not a plain FITS file.
+
+    valid_formats = []
+
+    for func, label, identifier, priority in data_factory:
+
         if func is auto_data:
             continue
-        if (func is default) and identifier(filename, **kwargs):
-            return func
 
-    # if that fails, try everything
-    for func, _, identifier in data_factory:
-        if func is auto_data:
+        try:
+            is_format = identifier(filename, **kwargs)
+        except ImportError:  # dependencies missing
             continue
-        if identifier(filename, **kwargs):
-            return func
 
+        if is_format:
+
+            valid_formats.append((priority, label, func))
+
+    priorities = list(zip(*valid_formats))[0]
+
+    highest_priority = max(priorities)
+
+    if priorities.count(highest_priority) > 1:
+        raise ValueError("Ambiguous file type")
```

Comment:
This seems too extreme. Perhaps we can choose an arbitrary factory and issue a warning?